001-resources-v2/proposal.md

```diff
@@ -0,0 +1,626 @@
+`from` value was never specified.
+
+The resource should determine a "default space", if any. Having a default space
+is useful for things like Git repos which have a default branch, or version
+spaces (e.g. `1.8`, `2.0`) which can point to the latest version line by
+default. If there is no default space, the user must specify it explicitly in
+the pipeline, either by configuring one on the resource (`default_space: foo`)
+or on every `get` step using the resource (`spaces: [foo]`).
+
+#### example
+
+Given the following request on `stdin`:
+
+```json
+{
+  "config": {
+    "uri": "https://github.com/concourse/concourse"
+  },
+  "from": {
+    "master": {"ref": "abc123"},
+    "feature/foo": {"ref":"def456"},
+    "feature/bar": {"ref":"987cia"}
+  },
+  "response_path": "/tmp/check-response.json"
+}
+```
+
+If the `feature/foo` branch has new commits, `master` is the default branch and
+has no new commits, and `feature/bar` has been `push -f`ed, you may see
+something like the following in `/tmp/check-response.json`:
+
+```json
+{"action":"discovered","space":"feature/foo","version":{"ref":"abcdf8"},"metadata":[{"name":"message","value":"fix thing"}]}
+{"action":"reset","space":"feature/bar"}
+{"action":"discovered","space":"feature/bar","version":{"ref":"abcde0"},"metadata":[{"name":"message","value":"initial commit"}]}
+{"action":"discovered","space":"feature/bar","version":{"ref":"abcde1"},"metadata":[{"name":"message","value":"add readme"}]}
+{"action":"default_space","space":"master"}
+{"action":"discovered","space":"feature/foo","version":{"ref":"abcdf9"},"metadata":[{"name":"message","value":"fix thing even more"}]}
+{"action":"discovered","space":"feature/bar","version":{"ref":"abcde2"},"metadata":[{"name":"message","value":"finish the feature"}]}
+```
+
+A few things to note:
+
+* A `reset` event is emitted immediately upon detecting that the given version
+  for `feature/bar` (`987cia`) is no longer available, followed by a
+  `discovered` event for every commit going back to the initial commit on the
+  branch.
+
+* No versions are emitted for `master`, because it's already up to date
+  (`abc123` is the latest commit).
+
+* The versions detected for `feature/foo` may appear between events for
+  `feature/bar`, as they're for unrelated spaces. The order only matters within
+  the space.
+
+
+### `get`: Fetch a version from the resource's space.
+
+The `get` command will be invoked with the following JSON structure on `stdin`:
+
+```go
+type GetRequest struct {
+  Config  Config  `json:"config"`
+  Space   Space   `json:"space"`
+  Version Version `json:"version"`
+}
+```
+
+The command will be invoked with a completely empty working directory. The
+command should populate this directory with the requested bits. The `git`
+resource, for example, would clone directly into the working directory.
+
+If the requested version is unavailable, the command should exit nonzero.
+
+No response is expected.
```

Comment:
After discussing with @clarafu we think it'd actually be better if `get`s still returned a response, and have that response include metadata.

This makes `get`, `check`, and `put` all have a response, and lets the resource author decide where metadata should be discovered. Now the difference between v1 and v2 would be that with v2 `check` can *also* return metadata - an additive change, rather than requiring all metadata come from `check`.

This way if there are situations where it's too expensive to collect all the metadata at once, it can be deferred until later `get` and `put` calls. Any metadata returned by `get` and `put` would just update the version's metadata.

This also makes the implementation in the ATC a bit cleaner because we don't have to worry about a scenario where e.g. a `get` is running with a version that has never had a `check` run and so there's no metadata to show to the user in the UI. This normally won't happen but is theoretically possible via the builds API, where you can submit a build that fetches an arbitrary version. This also makes v1 and v2 more similar to each other, minimizing conditional code paths in the implementation.